python/ray/air/tests/test_result.py

```diff
@@ -0,0 +1,101 @@
+import os
+import pytest
+
+import ray
+from ray.air import CheckpointConfig, RunConfig, ScalingConfig, session
+from ray.air.result import Result
+from ray.train.torch import TorchCheckpoint, TorchTrainer
+from ray.train.base_trainer import TrainingFailedError
+from ray.tune import TuneConfig, Tuner
+
+
+@pytest.fixture
+def ray_start_4_cpus():
+    address_info = ray.init(num_cpus=4)
+    yield address_info
+    # The code after the yield will run as teardown code.
+    ray.shutdown()
+
+
+def build_dummy_trainer(configs):
+    def worker_loop():
+        for i in range(configs["NUM_ITERATIONS"]):
+            session.report(
+                metrics={"metric_a": i, "metric_b": -i},
+                checkpoint=TorchCheckpoint.from_dict({"iter": i}),
+            )
+        raise RuntimeError()
+
+    trainer = TorchTrainer(
+        train_loop_per_worker=worker_loop,
+        scaling_config=ScalingConfig(num_workers=2, use_gpu=False),
+        run_config=RunConfig(
+            name=configs["EXP_NAME"],
+            storage_path=configs["STORAGE_PATH"],
+            checkpoint_config=CheckpointConfig(
+                num_to_keep=configs["NUM_CHECKPOINTS"],
+                checkpoint_score_attribute="metric_a",
+                checkpoint_score_order="max",
+            ),
+        ),
+    )
+    return trainer
+
+
+def build_dummy_tuner(configs):
+    return Tuner(
+        trainable=build_dummy_trainer(configs), tune_config=TuneConfig(num_samples=1)
+    )
+
+
+@pytest.mark.parametrize("mode", ["trainer", "tuner"])
+def test_result_restore(ray_start_4_cpus, tmpdir, mode):
+    NUM_ITERATIONS = 5
+    NUM_CHECKPOINTS = 3
+    storage_path = str(tmpdir)
+    exp_name = "test_result_restore"
+
+    configs = {
+        "EXP_NAME": exp_name,
+        "STORAGE_PATH": storage_path,
+        "NUM_ITERATIONS": NUM_ITERATIONS,
+        "NUM_CHECKPOINTS": NUM_CHECKPOINTS,
+    }
+
+    if mode == "trainer":
+        trainer = build_dummy_trainer(configs)
+        with pytest.raises(TrainingFailedError):
+            trainer.fit()
+    elif mode == "tuner":
+        tuner = build_dummy_tuner(configs)
+        tuner.fit()
+
+    # Find the trial directory to restore
+    exp_dir = os.path.join(storage_path, exp_name)
+    for dirname in os.listdir(exp_dir):
+        if dirname.startswith("TorchTrainer"):
+            trial_dir = os.path.join(exp_dir, dirname)
+            break
```

Comment:
Hmm, ideally we don't need to run Tune/Train and we could just mock the trial directory using the default loggers. But the Tune metadata thing gets created by Tune Trainable.save, which is a bit harder to mock.

I think this is ok for now, but wondering how we can improve testing to not do a full run each time. There's [this testing harness](https://github.com/ray-project/ray/blob/7c7361f93741e57149d571c4bdfe76c3e74c7953/python/ray/tune/tests/utils/experiment.py#L103) that kind of does what we want, but not exactly because it doesn't create real trial checkpoints.

Reply:
Yeah I was trying to use `result.path` at the beginning, but when we throw an error in the `worker_loop()`, the trainer fails before returning a result object, That's why I use `dirname.startswith("TorchTrainer")` to find the trial dir.

Yeah I agree, currently we have to start e2e training/tuning to generate a trial folder, which conflict with the idea of unit test. It'd be better to have a `generate_trial_folder()` mock function, which aligns with the behavior of real trainer & tuner.